host/sgx/sgxquote.c

```diff
@@ -125,18 +163,213 @@ static bool _load_sgx_dcap_ql(void)
+                    uuid = &_ecdsa_p256_uuid;
+                    local_mapped[i] = true;
+                    mapped_key_id_count++;
+                    break;
+                case SGX_QL_ALG_ECDSA_P384:
+                    uuid = &_ecdsa_p384_uuid;
+                    local_mapped[i] = true;
+                    mapped_key_id_count++;
+                    break;
+                default:
+                    uuid = &_unknown_uuid;
+                    local_mapped[i] = false;
+                    OE_TRACE_ERROR(
+                        "algorithm_id=%d maps to no uuid",
+                        key->base.algorithm_id);
+                    break;
+            }
+            memcpy(local_uuid + i, uuid, sizeof(oe_uuid_t));
+        }
+
+        _quote_ex_library.key_id_count = att_key_id_num;
+        _quote_ex_library.mapped_key_id_count = mapped_key_id_count;
+        _quote_ex_library.mapped = local_mapped;
+        _quote_ex_library.uuid = local_uuid;
+        _quote_ex_library.sgx_key_id = local_key_id;
+        local_mapped = NULL;
+        local_uuid = NULL;
+        local_key_id = NULL;
+
+        OE_TRACE_INFO(
+            "key_id_count=%lu mapped=%lu\n",
+            att_key_id_num,
+            mapped_key_id_count);
+
+        result = OE_OK;
+    }
+
+done:
+    if (local_mapped)
+    {
+        oe_free(local_mapped);
+        local_mapped = NULL;
+    }
+    if (local_uuid)
+    {
+        oe_free(local_uuid);
+        local_uuid = NULL;
+    }
+    if (local_key_id)
+    {
+        oe_free(local_key_id);
+        local_key_id = NULL;
+    }
+    if (_quote_ex_library.load_result == OE_OK)
+        _quote_ex_library.load_result = result;
+
+    OE_TRACE_INFO(
+        "_load_quote_ex_library_once() result=%s\n",
+        oe_result_str(_quote_ex_library.load_result));
+
+    return;
+}
+
+// For choosing between the DCAP library and the quote-ex library for quote
+// generation: the quote-ex library will be used if it can be loaded
+// and initialized successfully, otherwise the DCAP library is used.
+
+static bool _use_quote_ex_library(void)
+{
+    static oe_once_type once = OE_H_ONCE_INITIALIZER;
+    oe_once(&once, _load_quote_ex_library_once);
+
+    return (_quote_ex_library.load_result == OE_OK);
+}
+
 oe_result_t oe_sgx_qe_get_target_info(
     const oe_uuid_t* format_id,
     const void* opt_params,
     size_t opt_params_size,
     uint8_t* target_info)
 {
-    oe_result_t result = OE_FAILURE;
+    oe_result_t result = OE_UNEXPECTED;
     quote3_error_t err = SGX_QL_ERROR_UNEXPECTED;
 
-    OE_UNUSED(format_id);
-    OE_UNUSED(opt_params);
-    OE_UNUSED(opt_params_size);
+    if (!format_id || !target_info)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    if (_use_quote_ex_library())
+    {
+        sgx_status_t status = SGX_ERROR_UNEXPECTED;
+        sgx_att_key_id_ext_t updated_key_id = {{0}};
+        size_t tmp_size = 0;
+        uint8_t* tmp_buffer = NULL;
+        sgx_target_info_t tmp_target_info;
```

Comment:
nit: you updated the variables above to remove the abbreviation, but not throughout the rest of this file, such as here.

Reply:
Got it. Thanks!